File: frontend/src/api/feedback.ts
// フィードバック収集 API クライアント (Issue #79 / バックエンド #78)
//
// 認証基盤は未整備のため、org_id は暫定でローカル保存値を使い、X-Org-Id ヘッダで
// 送る（バックエンドは X-Org-Id == body.org_id を検証する）。認証導入時にここを
// トークン由来の org_id に差し替える。

const API_BASE = import.meta.env.VITE_API_BASE_URL
  ? `${import.meta.env.VITE_API_BASE_URL}/api`
  : '/api'

const ORG_STORAGE_KEY = 'meeting_score_org_id'
const DEFAULT_ORG_ID = 'org_local'

export function getOrgId(): string {
  try {
    return localStorage.getItem(ORG_STORAGE_KEY) || DEFAULT_ORG_ID
  } catch {
    // localStorage 不可（SSR / プライベートモード等）でも動く
    return DEFAULT_ORG_ID
  }
}

export function setOrgId(orgId: string): void {
  try {
    localStorage.setItem(ORG_STORAGE_KEY, orgId)
  } catch {
    // 保存できなくても致命的でないため握りつぶす
  }
}

async function postFeedback<T>(path: string, body: Record<string, unknown>): Promise<T> {
  const orgId = getOrgId()
  const res = await fetch(`${API_BASE}${path}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', 'X-Org-Id': orgId },
    body: JSON.stringify({ org_id: orgId, ...body }),
  })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  return res.json() as Promise<T>
}

export type PairwiseWinner = 'A' | 'B' | 'tie'
export type FlagDirection = 'overrated' | 'underrated'

export interface FeedbackAck {
  id: string
  generated_pairs: number
}

export interface FeedbackStats {
  org_id: string
  n_pairwise: number
  n_topk: number
  n_axis_flag: number
  stage: number
  next_stage: number | null
  pairs_to_next_stage: number | null
}

/** ペアワイズ比較を 1 件送る（手動ペア用）。 */
export const postPairwise = (params: {
  meeting_id: string
  utt_a: string
  utt_b: string
  winner: PairwiseWinner
}): Promise<FeedbackAck> =>
  postFeedback('/feedback/pairwise', { ...params, source: 'manual_pair' })

/**
 * Top5 訂正を送る。サーバ側で「Top入りした発言 × 押し出された発言」のペアワイズに
 * 自動展開される（generated_pairs に件数が返る）。
 */
export const postTopK = (params: {
  meeting_id: string
  corrected_top5: string[]
  original_top5: string[]
}): Promise<FeedbackAck> => postFeedback('/feedback/topk', params)

/** 発言カードの 👎（過大/過小評価 + 任意の軸 + コメント）を送る。 */
export const postAxisFlag = (params: {
  meeting_id: string
  utterance_id: string
  direction: FlagDirection
  axis?: string | null
  comment?: string | null
}): Promise<FeedbackAck> => postFeedback('/feedback/axis_flag', params)

/** 蓄積件数・段階(0/1/2)・次段階までの不足ペア数を取得する。 */
export const getFeedbackStats = async (): Promise<FeedbackStats> => {
  const orgId = getOrgId()
  const res = await fetch(
    `${API_BASE}/feedback/stats?org_id=${encodeURIComponent(orgId)}`,
    { headers: { 'X-Org-Id': orgId } },
  )
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  return res.json() as Promise<FeedbackStats>
}

File: backend/app/schemas/feedback.py
"""フィードバック収集 API の入出力スキーマ (Issue #78)

API 層の入力検証は ``Literal`` で厳密に行い、DB 層 (feedback_models) は
CHECK 制約で二重に守る。
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

Winner = Literal["A", "B", "tie"]
PairwiseSource = Literal["top5_reorder", "manual_pair"]
Direction = Literal["overrated", "underrated"]


class FeedbackPairwise(BaseModel):
    """POST /api/feedback/pairwise のリクエスト。"""

    org_id: str
    meeting_id: str
    utt_a: str
    utt_b: str
    winner: Winner
    source: PairwiseSource = "manual_pair"
    annotator: str | None = None


class FeedbackTopK(BaseModel):
    """POST /api/feedback/topk のリクエスト。

    ``corrected_top5`` は並べ替え後、``original_top5`` は元の Top5。
    サーバ側で差分からペアワイズを自動生成する。
    """

    org_id: str
    meeting_id: str
    corrected_top5: list[str]
    original_top5: list[str]
    annotator: str | None = None


class FeedbackAxisFlag(BaseModel):
    """POST /api/feedback/axis_flag のリクエスト。"""

    org_id: str
    meeting_id: str
    utterance_id: str
    direction: Direction
    axis: str | None = None
    comment: str | None = Field(default=None, max_length=80)
    annotator: str | None = None


class FeedbackAck(BaseModel):
    """書き込み系エンドポイントの共通レスポンス。"""

    id: str
    # topk から自動生成されたペアワイズ件数 (pairwise/axis_flag では 0)
    generated_pairs: int = 0


class FeedbackStats(BaseModel):
    """GET /api/feedback/stats のレスポンス。

    段階は Epic #77 の閾値に従う:
    - 段階 0: 0〜49 ペア (ベースモデル + デフォルト重み)
    - 段階 1: 50 ペア以上 (組織別重みプロファイル)
    - 段階 2: 300 ペア以上 (組織別 LoRA)
    """

    org_id: str
    n_pairwise: int  # feedback_pairwise の総数 (top5_reorder 展開分を含む)
    n_topk: int
    n_axis_flag: int
    stage: int  # 0 | 1 | 2
    next_stage: int | None  # 次の段階 (段階 2 到達後は None)
    pairs_to_next_stage: int | None  # 次段階までの不足ペア数 (段階 2 到達後は None)
